geomstats/stiefel.py

```diff
@@ -0,0 +1,351 @@
+        tangent_vec_a = gs.to_ndarray(tangent_vec_a, to_ndim=3)
+        n_tangent_vecs_a, _, _ = tangent_vec_a.shape
+
+        tangent_vec_b = gs.to_ndarray(tangent_vec_a, to_ndim=3)
+        n_tangent_vecs_b, _, _ = tangent_vec_b.shape
+
+        assert(n_tangent_vecs_a == n_tangent_vecs_b)
+
+        return gs.einsum("zij,zij->z", tangent_vec_a, tangent_vec_b)
+
+    def mean(self, points,
+             weights=None, n_max_iterations=32, epsilon=EPSILON):
+        """
+        Frechet mean of (weighted) points.
+        """
+        # TODO(nina): profile this code to study performance,
+        # i.e. what to do with sq_dists_between_iterates.
+
+        n_points = len(points)
+        assert n_points > 0
+
+        if weights is None:
+            weights = gs.ones(n_points)
+
+        n_weights = len(weights)
+        assert n_points == n_weights
+        sum_weights = gs.sum(weights)
+
+        mean = points[0]
+        if n_points == 1:
+            return mean
+
+        sq_dists_between_iterates = []
+        iteration = 0
+        while iteration < n_max_iterations:
+            a_tangent_vector = self.log(mean, mean)
+            tangent_mean = gs.zeros_like(a_tangent_vector)
+
+            for i in range(n_points):
+                # TODO(nina): abandon the for loop
+                point_i = points[i]
+                weight_i = weights[i]
+                tangent_mean = tangent_mean + weight_i * self.log(
+                                                    point=point_i,
+                                                    base_point=mean)
+            tangent_mean /= sum_weights
+
+            mean_next = self.exp(
+                tangent_vec=tangent_mean,
+                base_point=mean)
+
+            sq_dist = self.squared_dist(mean_next, mean)
+            sq_dists_between_iterates.append(sq_dist)
+
+            variance = self.variance(points=points,
+                                     weights=weights,
+                                     base_point=mean_next)
+            if gs.isclose(variance, 0):
+                break
+            if sq_dist <= epsilon * variance:
+                break
+
+            mean = mean_next
+            iteration += 1
+
+        if iteration is n_max_iterations:
+            print('Maximum number of iterations {} reached.'
+                  'The mean may be inaccurate'.format(n_max_iterations))
+        return mean
+
+    def exp(self, tangent_vec, base_point):
+        """
+        Riemannian exponential of a tangent vector wrt to a base point.
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
+        n_tangent_vecs, _, _ = tangent_vec.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_tangent_vecs == n_base_points
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        A = gs.matmul(gs.transpose(base_point, axes=(0, 2, 1)), tangent_vec)
```

Comment:
Ditto:
avoid uppercases letter for variable names. avoid single letter for variable names.